### ISSUE TITLE
[feature request] please include a OAuth2 token based authentication

### ISSUE THREAD

**Issue by trsh** [cutelyst/cutelyst#126]

**Comment by trsh** [comment]
@dantti 

In mean while. I have a situation. My front-end, that makes requests is on domain&server X, and Cutelyst Api on domain&server Y. So the cookies are not sent, and I can not use this authorization. 

Maybe there is some header I can add, to submit the cookie? Or the only solution here is to send over user and pass every time?

**Comment by dantti** [comment]
If I understood correctly you could try this:
https://stackoverflow.com/questions/3342140/cross-domain-cookies
I haven't done this myself...

**Comment by dantti** [comment]
awesome :+1:

**Comment by trsh** [comment]
@dantti 

any ideas on these:

`error: expected unqualified-id before ‘)’ token
 Q_LOGGING_CATEGORY(OAUTH2SERVER_H, "cutelyst.plugin.authentication")
`
? Im stuck

**Comment by dantti** [comment]
I believe OAUTH2SERVER_H is defined at the header, so use another identifier

**Comment by trsh** [comment]
@dantti 

Can you move all crypto stuff into some common class? So that stuff that remains in credentialpassword(_p).h is like:

```
  OAuth2User authenticate(Context *c, OAuth2Realm *realm, const ParamsMultiMap &authinfo);

  QString passwordField() const;
  void setPasswordField(const QString &fieldName);
  PasswordType passwordType() const;
  void setPasswordType(PasswordType type);
  QString passwordPreSalt() const;
  void setPasswordPreSalt(const QString &passwordPreSalt);
  QString passwordPostSalt() const;
  void setPasswordPostSalt(const QString &passwordPostSalt);
  static bool validatePassword(const QByteArray &password, const QByteArray &correctHash);

..

bool checkPassword(const OAuth2User &user, const ParamsMultiMap &authinfo);

```
Because all other things will be also needed by OAuth2 plugin. And I don't want it to be dependent on current authentication plugin.

**Comment by dantti** [comment]
Hmm I will think about that but since v2 is ABI stable I am not sure I can, at least the way I am thinking of

**Comment by trsh** [comment]
@dantti didn't get the problem! o_O .. so it will be 2.1?

**Comment by dantti** [comment]
Changing ABI requires v3, also OAuth2 would be a lot more useful if it were just another backend to the Authentication module.

**Comment by trsh** [comment]
@dantti 

Sorry but I disagree. 

In `OAuth2` lot's of stuff from `Authentication` is redundant - like there is no need for cookies and session, and more. So users who would like to use `OAuth2` only (typical for REST api usage), would have a lot of unused boilerplate.  At beginning I tried to mix them, but droped the idea very fast. Actually only thing that was Really in-common is this password generation and check. 

It would be nice, if password common stuff is bundled like an sub-plugin/lib, that is added via cmake. This way we can just forward the methods and think about breaking changes in v3. 

What do you think?

**Comment by dantti** [comment]
IMO at the moment if you can just copy the copy as it's only a few functions it would be ok,
later when we have something working we can check on this.
I want to benchmark some others PBKDF2 implementations as they might be faster so
maybe a plugin based PBKDF2-Qt lib could be done.

**Comment by trsh** [comment]
Deal!

**Comment by dantti** [comment]
Are you still working on this? I might need this feature for a possible project... :smile:

**Comment by trsh** [comment]
@dantti 

My work was like 85% finished, when I was thrown in to some other project. I could finish it by end of June when it ends. Or send you now a .zip with code so far. 

Note that it doesn't include (and I won't add) authorization with redirect URL and authorization string, as I don't need it. Just these grant types:

```
#define GRANT_TYPE_CLIENT_CREDENTIALS QStringLiteral("client_credentials")
#define GRANT_TYPE_PASSWORD QStringLiteral("password")
#define GRANT_TYPE_REFRESH_TOKEN QStringLiteral("refresh_token")
```

and `authenticate` for token check.

**Comment by dantti** [comment]
Cool, yeah I also have other stuff consuming my time ATM but I'm slowly studying how this works.

I have created a oauth2 branch can you do a PR to that branch so that we can keep you in history?

**Comment by trsh** [comment]
@dantti 

the problem is, that I do not have it at this moment, in cutelyst sources. It's in my local project under some folder plugins. It seemed faster to dev this way, and the transition was planed for later.

So again it's a ZIP or wait until end of June.

Let me know.

**Comment by dantti** [comment]
dantti12->gmail, it's in the headers of files :P

**Comment by trsh** [comment]
@dantti done. Check mail. Sorry for 6 day delay - at some moment it totally jumped out of my head.

**Comment by sanekt197** [comment]
dantti12, Do not share the implementation of OAuth please? :) Also came to the need for OAuth. It would be nice to have an API

**Comment by dantti** [comment]
What I have from the email is work in progress that I didn't had the time to review yet, too busy ATM...

**Comment by mbnoimi** [comment]
Hi guys,
Is there any progress on this issue? I can see the last commit in OAuth2 branch done since two years ago.

**Comment by dantti** [comment]
nope, I have talked to OAuth2 servers and each of them work quite differently, not sure I can come up with something generic.
